### PR TITLE
Amqp hard soft limits

### PR DESCRIPTION
**Pull request by mtaylor**
For discussion purposes.  Do not merge.

Closing the coordinator link on reject presettled tx is described in spec here: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transactions-v1.0-os.html#doc-idp145616

### CODE REVIEW

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/settings/impl/AddressSettings.java [hunk 1]
```diff
@@ -549,7 +569,8 @@ public int getEncodeSize() {
          BufferHelper.sizeOfNullableBoolean(autoDeleteJmsQueues) +
          BufferHelper.sizeOfNullableBoolean(autoCreateJmsTopics) +
          BufferHelper.sizeOfNullableBoolean(autoDeleteJmsTopics) +
-         BufferHelper.sizeOfNullableInteger(managementBrowsePageSize);
+         BufferHelper.sizeOfNullableInteger(managementBrowsePageSize) +
+         BufferHelper.sizeOfNullableLong(maxSizeBytesRejectThreshold);
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/settings/impl/AddressSettings.java]: I suspect this could break compatibility on the bindings journal? better to check.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/proton/ProtonTest.java [hunk 2]
```diff
@@ -234,42 +235,6 @@ public void testTemporaryQueue() throws Throwable {
    }
 
    @Test
-   public void testCommitProducer() throws Throwable {
```
Comment (clebertsuconic) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/proton/ProtonTest.java]: Isn't this a test added by Andy recently? why removed it?

Comment (mtaylor) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/proton/ProtonTest.java]: Yup. I need to clean up the commits. This shouldn't be removed.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/settings/impl/AddressSettings.java (hunk 1 above)
Comment (mtaylor) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/settings/impl/AddressSettings.java]: Possibly.  We've done something similar re e try with auto delete queues so Im not too worried.  I can test this.

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/proton/plug/ProtonSessionIntegrationCallback.java [hunk 3]
```diff
@@ -352,17 +354,34 @@ public void serverSend(final Receiver receiver,
...
       if (store.isRejectingMessages()) {
-         ErrorCondition ec = new ErrorCondition(AmqpError.RESOURCE_LIMIT_EXCEEDED, "Address is full: " + message.getAddress());
-         Rejected rejected = new Rejected();
-         rejected.setError(ec);
-         delivery.disposition(rejected);
-         connection.flush();
+         // We drop pre-settled messages (and abort any associated Tx)
+         if (delivery.remotelySettled()) {
+            if (serverSession.getCurrentTransaction() != null) {
+               rollbackCurrentTX(false);
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/proton/plug/ProtonSessionIntegrationCallback.java]: There's a thing you can do here.. markRollbackOnly on the current Transaction?

**Comment by clebertsuconic** [comment]
It looks good... only thing there's a test failing (that's part of the fast-tests.. the PR build is accurate)

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/proton/plug/ProtonSessionIntegrationCallback.java [hunk 4]
```diff
@@ -351,18 +353,35 @@ public void serverSend(final Receiver receiver,
...
-         rejected.setError(ec);
-         delivery.disposition(rejected);
-         connection.flush();
+      if (store.isRejectingMessages()) {
+         // We drop pre-settled messages (and abort any associated Tx)
+         if (delivery.remotelySettled()) {
+            if (serverSession.getCurrentTransaction() != null) {
+               rollbackCurrentTX(false);
+
+               // Spec requires closure of coordinator link on rejection of presettled messages within a tx
```
Comment (gemmellr) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/proton/plug/ProtonSessionIntegrationCallback.java]: It requires either closing the coordinator link, or that you reject attempts to discharge the transaction that don't have fail=true (i.e fail commit attempts, allow rollback attempts).

I'd personally go for the second option as its the simpler for clients/applications to deal with, and since the use of pre-settled transfers mean its highly possible the client will have already sent everything and attempted to commit the transaction before it ever sees the coordinator link closure so there will be little difference between the two options from the brokers perspective.

Comment (mtaylor) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/proton/plug/ProtonSessionIntegrationCallback.java]: Updating here, so we can track the conversation.  There spec outlines 2 options:  If a pre-settled message is sent during a tx that the broker can't handle it should either (in laymans terms):

1. Close the coordinator link.
2. Set the transaction to rollback only.

Given the 2 options I went with 1. Close the coordinator link.  

The benefit of 1. is that the client knows the Tx failed early on and can stop processing it.  Robbie mentioned that a long running tx is unlikely and probably the client already sent everything + the commit.

The benefit of 2. is that the client does not need to re establish the coordinator link if the tx fails in this scenario.  Robbie mentions that clients may support this option.

Comment (mtaylor) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/proton/plug/ProtonSessionIntegrationCallback.java]: My opinion is that, this fix is spec compliant and fixes the intended issue.  I'd prefer to merge this and close out the hard/soft limit issue.  If people prefer us to use option 2. then we can create another issue to cover the switch in policy, any client that does not handle this scenario is surely not spec compliant?

Comment (gemmellr) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/proton/plug/ProtonSessionIntegrationCallback.java]: The change is spec compliant, certainly, so feel free to merge away if you want. I was mainly pointing out the comment is imprecise and that of the options available I think its the less friendly to go with and less likely to just work.

Clients/applications still cant tell message(s) might be of specific interest, can't tell which transaction caused a problem if there are multiple in progress on the link (not that its necessarily currently supported with Artemis), and have to handle the coordinator link closure and setting up a new link etc to continue working when using this option, versus the other option where they just handle the commit success/failure that they would always have to anyway. Its true that from a time perspective you may find out earlier but in a lot of cases I don't think that will be significant, even where it is a long running transaction (since its fairly likely the process that takes a long time happens before the send that causes an issue).

You're right that it can be said that any clients / applications (depending on where the exact responsibility for handing the scenario lands in a given implementation) that don't deal with the scenario well simply aren't implementing things that the spec says are possible and they should do so.

Comment (mtaylor) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/proton/plug/ProtonSessionIntegrationCallback.java]: @gemmellr The comment is there just to show why the method behaves in this way, I didn't add all options.  But, given your feedback, I think it's worth changing this to option 2.  I'll update the PR.

Comment (clebertsuconic) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/proton/plug/ProtonSessionIntegrationCallback.java]: It's how everything else works on Artemis in regards to rollbacks and TX.

**Comment by clebertsuconic** [comment]
This is ready to merge IMO. Tests are good.

Only thing missing is.. I don't see a JIRA-id on any of these commits? shoulnd't that be added here?
